src/utils.ts

```diff
@@ -1,76 +1,103 @@
+  const key = createCID(buffer, multicode)
+  return [key, buffer]
+}
+
+// TODO: implement a streaming hashing function for large files.
+// Note: in fact this returns a serialized CID, not a CID object.
+export function createCID (data: string | Uint8Array, multicode = multicodes.RAW): string {
+  const uint8array = typeof data === 'string' ? new TextEncoder().encode(data) : data
+  const digest = multihasher.digest(uint8array)
+  return CID.create(1, multicode, digest).toString(multibase.encoder)
+}
+
 export function exit (message: string): never {
   console.error('[chel]', colors.red('Error:'), message)
   Deno.exit(1)
 }
 
-export async function getBackend (src: string): Promise<Backend> {
-  let from
-  let options
-  if (isDir(src)) {
-    from = 'fs'
-    options = { internal: true, dirname: src }
-  } else if (isFile(src)) {
-    from = 'sqlite'
-    options = { internal: true, dirname: path.dirname(src), filename: path.basename(src) }
-  } else throw new Error(`invalid argument: "${src}"`)
+// Supported backend types: fs and sqlite
+export async function getBackend (src: string, { type, create } = { type: '', create: false }): Promise<Backend> {
+  const fsOptions = { internal: true, dirname: src }
+  const sqliteOptions = { internal: true, dirname: path.dirname(src), filename: path.basename(src) }
+
+  // If 'create' is falsy then make sure 'src' already exists, becqause otherwise 'initStorage()' would create it.
+  if (!create && !await isDir(src) && !await isFile(src)) throw new Error(`not found: "${src}"`)
+
+  // We first need to know the backend type and init options.
+  let from = type
+  if (!from) {
+    // Attempt to auto-detect the backend type.
+    if (await isDir(src)) from = 'fs'
+    else if (await isFile(src)) from = 'sqlite'
+    else throw new Error(`could not infer backend type. Not found: "${src}"`)
+  }
+
+  let initOptions
+  switch (from) {
+    case 'fs': initOptions = fsOptions; break
+    case 'sqlite': initOptions = sqliteOptions; break
+    default: throw new Error(`unknown backend type: "${from}"`)
+  }
 
   const backend: Backend = backends[from as keyof typeof backends]
   try {
-    await backend.initStorage(options)
+    await backend.initStorage(initOptions)
   } catch (error) {
-    throw new Error(`could not init storage backend at "${src}": ${error.message}`)
+    throw new Error(`could not init '${from}' storage backend at "${src}": ${error.message}`)
   }
   return backend
 }
 
+export function getPathExtension (path: string): string {
+  const index = path.lastIndexOf(".")
+  if (index === -1 || index === 0) return ""
+  return path.slice(index).toLowerCase()
+}
+
 export function isArrayLength (arg: number): boolean {
   return Number.isInteger(arg) && arg >= 0 && arg <= 2 ** 32 - 1
 }
 
 // Checks whether a path points to a directory, following symlinks if any.
-export function isDir (path: string | URL): boolean {
+export async function isDir (path: string | URL): Promise<boolean> {
```

Comment:
Why did you change these to be async?

And if there's a strong reason to, can you add that as a comment?

Reply:
Nowadays async is generally the default way for any I/O operation in JavaScript, and the fact these functions were sync caught my eye. Agreed, not sure whether there is a real performance benefit in the CLI tool use case, at least as long as the functions aren't called in loops. But in principe it still allows the engine to not forcibly stay idle while waiting for the operation to complete. And we might start using these functions in loops sooner or later, in which case being async enables usage of e.g. `Promise.all()`. 
Also these function names have no `-Sync` suffix unlike native Nodejs and Deno sync IO functions

Reply:
Alright 👍